packages/kokkos/core/src/impl/Kokkos_Error.cpp

```diff
@@ -56,6 +56,28 @@
 
 namespace Kokkos {
 namespace Impl {
+static void (*violation_handler_pointer)(violation_info const&) = default_violation_handler;
+}
+
+void set_violation_handler(void (*new_handler)(violation_info const&)) {
+  violation_handler_pointer = new_handler;
+}
+
+namespace Impl {
+
+void default_violation_handler(violation_info const& info) {
+  fprintf(stderr, "contract violation in %s:%d\nfunction %s\n%s\n",
+      info.file_name, info.line, info.function_name, info.comment);
+  std::stringstream string_stream;
+  traceback_callstack(string_stream);
+  std::string s = string_stream.str();
+  fprintf(stderr, "%s\n", s.c_str());
```

Comment:
Is this function run on the device or always on the host?  If so, whe using fprint() and not std::cerr?

Reply:
(1) This function always runs on the host, (2) I use `fprintf` when reporting errors in an MPI-parallel setting because it is usually better at keeping messages from different ranks from getting mixed up with one another. It is possible that a single `operator<<` call to `std::cerr` might give the same atomicity, but I did this out of habit.

Reply:
@ibaned wrote:
> It is possible that a single `operator<<` call to `std::cerr` might give the same atomicity, but I did this out of habit.

I have good luck first building a string (with `ostringstream`) and then writing it once to the `ostream`.  I'm perfectly fine with using `fprintf` here for this reason.

Reply:
Yup, that is the trick.  You build a `std::ostringstream` first then output the entire string in one `std::cerr <<` operator invocation.